server/src/streaming/systems/snapshot.rs

```diff
@@ -0,0 +1,76 @@
+use crate::streaming::session::Session;
+use crate::streaming::systems::system::System;
+use iggy::error::IggyError;
+use iggy::models::snapshot::Snapshot;
+use iggy::snapshot::{SnapshotCommand, SnapshotCompression};
+use std::io::Cursor;
+use std::io::Write;
+use tokio::process::Command;
+use tracing::error;
+use zip::write::{SimpleFileOptions, ZipWriter};
+
+impl System {
+    pub async fn get_snapshot(
+        &self,
+        session: &Session,
+        compression: SnapshotCompression,
+        snapshot_configs: Vec<SnapshotCommand>,
+    ) -> Result<Snapshot, IggyError> {
+        self.ensure_authenticated(session)?;
+        get_zip_data(compression, snapshot_configs).await
+    }
+}
+
+async fn get_zip_data(
+    compression: SnapshotCompression,
+    snapshot_commands: Vec<SnapshotCommand>,
+) -> Result<Snapshot, IggyError> {
+    let mut zip_buffer = Cursor::new(Vec::new());
+    let mut zip_writer = ZipWriter::new(&mut zip_buffer);
+
+    let compression = match compression {
+        SnapshotCompression::Stored => zip::CompressionMethod::Stored,
+        SnapshotCompression::Deflated => zip::CompressionMethod::Deflated,
+        SnapshotCompression::Bzip2 => zip::CompressionMethod::Bzip2,
+        SnapshotCompression::Lzma => zip::CompressionMethod::Lzma,
+        SnapshotCompression::Xz => zip::CompressionMethod::Xz,
+        SnapshotCompression::Zstd => zip::CompressionMethod::Zstd,
+    };
+
+    for command in snapshot_commands {
+        match Command::new(command.command)
+            .args(&command.args)
+            .output()
+            .await
+        {
+            Ok(out) => {
+                if let Err(e) = zip_writer.start_file(
+                    command.filename,
+                    SimpleFileOptions::default().compression_method(compression),
+                ) {
+                    error!("Failed to create snapshot file {}: {}", command.filename, e);
+                    continue;
+                }
+
+                if let Err(e) = zip_writer.write_all(&out.stdout) {
+                    error!(
+                        "Failed to write to snapshot file {}: {}",
+                        command.filename, e
+                    );
+                    continue;
+                }
+            }
+            Err(e) => {
+                error!("Failed to execute command {}: {}", command.command, e);
+                continue;
+            }
+        }
+    }
+
+    zip_writer
+        .finish()
+        .map_err(|_| IggyError::SnapshotFileCompletionFailed)?;
+    let zip_data = zip_buffer.into_inner();
+
```

Comment:
Is the snapshot file automatically deleted?

Reply:
hmm, the snapshot data is stored in-memory in zip_buffer (a Cursor<Vec<u8>>), rather than as a file on disk. Therefore, when zip_buffer goes out of scope at the end of this function, it will automatically be deleted from memory

Reply:
Ok, sounds good :)